libwnbd/wnbd_ioctl.cpp

```diff
@@ -133,6 +137,315 @@ DWORD WnbdOpenAdapterCMDeviceInstance(PDEVINST DeviceInstance)
+        Status = ERROR_NOT_ENOUGH_MEMORY;
+        LogError(
+            "calloc Error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+        goto exit;
+    }
+    Status = WnbdList(TempList, &BufferSize);
+    if (Status) {
+        LogError(
+            "WnbdList Error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+        goto exit;
+    }
+    CHAR* InstanceName;
+    for (ULONG index = 0; index < TempList->Count; index++) {
+        InstanceName = TempList->Connections[index].Properties.InstanceName;
+        WNBD_REMOVE_OPTIONS RemoveOptions = { 0 };
+        RemoveOptions.Flags.HardRemove = TRUE;
+        Status = WnbdRemoveEx(InstanceName, &RemoveOptions);
+        if (Status) {
+            LogError(
+                "Could not remove device: %s. "
+                "Error: %d. Error message: %s",
+                InstanceName, Status, win32_strerror(Status).c_str());
+            goto exit;
+        }
+    }
+
+exit:
+    if (TempList) {
+        free(TempList);
+    }
+    return Status;
+}
+
+DWORD WnbdUninstallDriver(PBOOL RebootRequired)
+{
+    DWORD Status;
+    HDEVINFO AdapterDevHandle = INVALID_HANDLE_VALUE;
+    SP_DEVINFO_DATA DevInfoData;
+    DevInfoData.cbSize = sizeof DevInfoData;
+    Status = FindDevice(&AdapterDevHandle, &DevInfoData);
+    if (ERROR_SUCCESS != Status) {
+        LogError("Driver not installed");
+        return Status;
+    }
+
+    
+    PWNBD_CONNECTION_LIST ConnList = NULL;
+    DWORD err = HardRemoveAllDevices();
+    if (err) {
+        return err;
+    }
+    
+    Status = WnbdRemoveDriver(AdapterDevHandle, &DevInfoData, RebootRequired);
+
+    if (INVALID_HANDLE_VALUE != AdapterDevHandle) {
+        SetupDiDestroyDeviceInfoList(AdapterDevHandle);
+    }
+
+    if (ERROR_SUCCESS != Status) {
+        LogError(
+            "Error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+    }
+
+    return Status;
+}
+
+DWORD InstallDriver(CHAR* FileNameBuf, HDEVINFO* AdapterDevHandle, PBOOL RebootRequired)
+{
+    CHAR ClassName[MAX_CLASS_NAME_LEN];
+    GUID ClassGuid;
+    SP_DEVINFO_DATA DevInfoData;
+    SP_DEVINSTALL_PARAMS_A InstallParams;
+    DWORD Status = 0;
+
+    if (!SetupDiGetINFClassA(FileNameBuf, &ClassGuid, ClassName, MAX_CLASS_NAME_LEN, 0)) {
+        Status = GetLastError();
+        goto exit;
+    }
+
+    if (INVALID_HANDLE_VALUE == (*AdapterDevHandle = SetupDiCreateDeviceInfoList(&ClassGuid, 0))) {
+        Status = GetLastError();
+        goto exit;
+    }
+
+    DevInfoData.cbSize = sizeof DevInfoData;
+    if (!SetupDiCreateDeviceInfoA(*AdapterDevHandle, ClassName, &ClassGuid, 0, 0, DICD_GENERATE_ID, &DevInfoData)) {
+        Status = GetLastError();
+        goto exit;
+    }
+
+    if (!SetupDiSetDeviceRegistryPropertyA(*AdapterDevHandle, &DevInfoData,
+                                           SPDRP_HARDWAREID, (PBYTE)WNBD_HARDWAREID, WNBD_HARDWAREID_LEN)) {
+        Status = GetLastError();
```

Comment:
Should we log all those errors? We might need to know which of those calls failed, not just InstallDriver itself. It might be useful for debugging purposes. Same applies to other errors throughout the PR.